File: yk/nand2tetris/06/asm.py
#!/usr/bin/env python3
import collections
import re
import sys


class Parser:
    def __init__(self):
        self._n_current = 0

        # コマンドライン引数をファイルパスと見なして読み込む。
        with open(sys.argv[1], "r") as f:
            # 行を要素とするリストとして保持する。
            self._lines = f.read().splitlines()

        for i in range(len(self._lines)):  # 各行について。
            line = self._lines[i]
            line = re.sub("//.*", "", line)  # コメントを削除する。
            line = line.strip()  # 左右の空白を削除する。
            self._lines[i] = line

        self._lines = [x for x in self._lines if x != ""]  # 空行を削除する。

    def _line(self):
        """現在の行の文字列を返す。"""
        return self._lines[self._n_current]

    def _split(self):
        line = self._line()
        dest = "null"
        jump = "null"
        ss = line.split("=")  # 「=」で分割したリストを得る。

        # 文字列に「=」が含まれるならば、左側を dest、右側を line とする。
        # （そうでなければ、dest は null、line は文字列全体である。）
        if len(ss) == 2:  # リストの要素数が 2 である。
            dest = ss[0]
            line = ss[1]

        ss = line.split(";")

        # line に「;」が含まれるならば、左側を line、右側を jump とする。
        # （そうでなければ、line はそのまま、jump は null である。）
        if len(ss) == 2:
            line = ss[0]
            jump = ss[1]

        comp = line  # line には今や comp ニーモニックだけが残っているはずである。
        return dest, comp, jump

    def hasMoreCommands(self):
        return self._n_current < len(self._lines)

    def advance(self):
        self._n_current += 1

    def commandType(self):
        line = self._line()

        if re.search("@.*", line):  # 「@」で始まる。
            return "A_COMMAND"

        if re.search("\(.*\)", line):  # 丸括弧で囲まれている。
            return "L_COMMAND"

        return "C_COMMAND"

    def symbol(self):
        line = self._line()
        type = self.commandType()

        if type == "A_COMMAND":
            return line[1:]  # 最初の文字「@」を除いて返す。

        if type == "L_COMMAND":
            return line[1:-1]  # 左右の丸括弧を除いて返す。

    def dest(self):
        dest, _, _ = self._split()
        return dest

    def comp(self):
        _, comp, _ = self._split()
        return comp

    def jump(self):
        _, _, jump = self._split()
        return jump


class Code:
    def __init__(self):
        mnemonic_dest = "null M D MD A AM AD AMD".split()
        binary_dest = "000 001 010 011 100 101 110 111".split()
        # リストの要素数が一致することを確認しておく。
        assert len(mnemonic_dest) == len(binary_dest)
        # ニーモニック -> バイナリコード、な連想配列を定義する。
        self._dict_dest = collections.OrderedDict(
            zip(mnemonic_dest, binary_dest))

        # compニーモニックについて、同様に連想配列を定める。
        mnemonic_comp = (
            "0 1 -1 D A !D !A -D -A D+1 A+1 D-1 A-1 D+A D-A A-D D&A D|A "
            "_ _ _ _ M _ !M _ -M  _ M+1 _ M-1 D+M D-M M-D D&M D|M").split()
        binary_comp = (
            "0101010 0111111 0111010 0001100 0110000 0001101 0110001 0001111 "
            "0110011 0011111 0110111 0001110 0110010 0000010 0010011 0000111 "
            "0000000 0010101 1101010 1111111 1111010 1001100 1110000 1001101 "
            "1110001 1001111 1110011 1011111 1110111 1001110 1110010 1000010 "
            "1010011 1000111 1000000 1010101").split()
        assert len(mnemonic_comp) == len(binary_comp)
        self._dict_comp = collections.OrderedDict(
            zip(mnemonic_comp, binary_comp))

        # jumpニーモニックについて、同様に連想配列を定める。
        mnemonic_jump = "null JGT JEQ JGE JLT JNE JLE JMP".split()
        binary_jump = "000 001 010 011 100 101 110 111".split()
        assert len(mnemonic_jump) == len(binary_jump)
        self._dict_jump = collections.OrderedDict(
            zip(mnemonic_jump, binary_jump))

    def dest(self, mnemonic):
        return self._dict_dest[mnemonic]

    def comp(self, mnemonic):
        return self._dict_comp[mnemonic]

    def jump(self, mnemonic):
        return self._dict_jump[mnemonic]


class SymbolTable:
    def __init__(self):
        # 連想配列を用意する。シンボル -> アドレス、として用いる。
        self._dict = collections.OrderedDict()

        # 定義済みシンボルを登録していく。
        self.addEntry("SP", 0)
        self.addEntry("LCL", 1)
        self.addEntry("ARG", 2)
        self.addEntry("THIS", 3)
        self.addEntry("THAT", 4)

        for i in range(0, 16):  # R0〜R15 を登録する。
            self.addEntry("R{}".format(i), i)

        self.addEntry("SCREEN", 16384)
        self.addEntry("KBD", 24576)

    def __str__(self):
        return str(self._dict)

    def addEntry(self, symbol, address):
        self._dict[symbol] = address

    def contains(self, symbol):
        return symbol in self._dict

    def getAddress(self, symbol):
        return self._dict[symbol]


def main():
    code = Code()
    table = SymbolTable()
    parser = Parser()
    address = 0

    # 最初のパスを行う。
    while parser.hasMoreCommands():
        type = parser.commandType()

        # ラベルシンボル (Xxx) をシンボルテーブルに登録する。
        if type == "L_COMMAND":
            symbol = parser.symbol()
            table.addEntry(symbol, address)

        # ROMアドレスを動かす。
        if type == "A_COMMAND" or type == "C_COMMAND":
            address += 1

        parser.advance()

    parser = Parser()
    address = 16  # 変数シンボル @Xxx のアドレスは 16 からである。

    # 2 回目のパスを行う。
    while parser.hasMoreCommands():
        type = parser.commandType()

        if type == "A_COMMAND":
            symbol = parser.symbol()

            if re.search("[^0-9]+", symbol):  # @Xxx が数字でない。
                if not table.contains(symbol):  # 未登録である。
                    # 変数シンボル @Xxx と登録する。
                    table.addEntry(symbol, address)
                    address += 1

                integer = table.getAddress(symbol)
            else:  # @Xxx が数字である。
                integer = int(symbol)  # "123" -> 123 などと型変換する。

            # 「0」でパディングしつつ二進数文字列を出力する。
            print("0{:015b}".format(integer))
        elif type == "C_COMMAND":
            # dest、comp、jump の各部分を二進数に変換して出力する。
            mnemonic_dest = parser.dest()
            binary_dest = code.dest(mnemonic_dest)
            mnemonic_comp = parser.comp()
            binary_comp = code.comp(mnemonic_comp)
            mnemonic_jump = parser.jump()
            binary_jump = code.jump(mnemonic_jump)
            print("111{}{}{}".format(binary_comp, binary_dest, binary_jump))

        parser.advance()


if __name__ == "__main__":
    main()

File: yk/nand2tetris/06/version_cpp/asm.cpp
#include <algorithm>
#include <bitset>
#include <cassert>
#include <fstream>
#include <iostream>
#include <map>
#include <string>
using namespace std;


enum CommandType {A_COMMAND, C_COMMAND, L_COMMAND};


class Parser {
private:
    ifstream stream;
    string line_current;
    string line_next;
    string s_dest;
    string s_comp;
    string s_jump;

    void split() {
        string s = line_current;
        s_dest = "null";
        s_jump = "null";
        size_t i = s.find("=");

        // 文字列に「=」が含まれるならば、左側を s_dest、右側を s とする。
        // （そうでなければ、s_dest は null、s は文字列全体である。）
        if (i != string::npos) {
            s_dest = s.substr(0, i);
            s = s.substr(i + 1);
        }

        i = s.find(";");

        // s に「;」が含まれるならば、左側を s、右側を s_jump とする。
        // （そうでなければ、s_jump は null、s はそのまま、である。）
        if (i != string::npos) {
            s_jump = s.substr(i + 1);
            s = s.substr(0, i);
        }

        s_comp = s;
    }
public:
    Parser(const char *path_input) {
        stream.open(path_input);
        assert(!stream.fail());  // ファイルが正常に開けなかった場合。
        advance();  // line_next に最初の行を得る。
    }

    ~Parser() {
        stream.close();
    }

    bool hasMoreCommands() {
        return line_next != "EOF";
    }

    void advance() {
        line_current = line_next;

        do {
            if (stream.peek() == EOF) {  // これ以上読み込むデータがない。
                line_next = "EOF";
            } else {
                getline(stream, line_next);  // 次の行を読み込む。
                // 改行文字がある場合は無害化しておく。
                replace(line_next.begin(), line_next.end(), '\r', ' ');
                size_t i = line_next.find("//");

                if (i == string::npos) {  // コメントを削除する。
                    ;
                } else {
                    line_next = line_next.substr(0, i);
                }

                i = line_next.find_first_not_of(' ');

                if (i == string::npos) {  // 左側の空白を削除する。
                    line_next = "";
                } else {
                    line_next = line_next.substr(i);
                }

                i = line_next.find_last_not_of(' ');

                if (i == string::npos) {  // 右側の空白を削除する。
                    line_next = "";
                } else {
                    line_next = line_next.substr(0, i + 1);
                }
            }
        } while (line_next == "");  // 空行は無視する。

        split();
    }

    CommandType commandType() {
        if (line_current[0] == '@') {  // 「@」で始まる。
            return A_COMMAND;
        } else if (line_current[0] == '(') {  // 「(」で始まる。
            return L_COMMAND;
        } else {
            return C_COMMAND;
        }
    }

    string symbol() {
        CommandType t = commandType();
        if (t == A_COMMAND) {  // 最初の文字「@」を除いて返す。
            return line_current.substr(1);
        } else if (t == L_COMMAND) {  // 左右の丸括弧を除いて返す。
            return line_current.substr(0, line_current.size() - 1).substr(1);
        } else {
            assert(false);
        }
    }

    string dest() { return s_dest; }
    string comp() { return s_comp; }
    string jump() { return s_jump; }
};


class Code {
private:
    // # ニーモニック -> バイナリコード、な連想配列を定義する。
    map<string, string> map_dest {
        {"null", "000"}, {"M", "001"}, {"D", "010"}, {"MD", "011"},
        {"A", "100"}, {"AM", "101"}, {"AD", "110"}, {"AMD", "111"}};
    map<string, string> map_comp {
        {"0", "0101010"}, {"1", "0111111"},
        {"-1", "0111010"}, {"D", "0001100"},
        {"A", "0110000"}, {"!D", "0001101"},
        {"!A", "0110001"}, {"-D", "0001111"},
        {"-A", "0110011"}, {"D+1", "0011111"},
        {"A+1", "0110111"}, {"D-1", "0001110"},
        {"A-1", "0110010"}, {"D+A", "0000010"},
        {"D-A", "0010011"}, {"A-D", "0000111"},
        {"D&A", "0000000"}, {"D|A", "0010101"},
        {"_", "1101010"}, {"_", "1111111"},
        {"_", "1111010"}, {"_", "1001100"},
        {"M", "1110000"}, {"_", "1001101"},
        {"!M", "1110001"}, {"_", "1001111"},
        {"-M", "1110011"}, {"_", "1011111"},
        {"M+1", "1110111"}, {"_", "1001110"},
        {"M-1", "1110010"}, {"D+M", "1000010"},
        {"D-M", "1010011"}, {"M-D", "1000111"},
        {"D&M", "1000000"}, {"D|M", "1010101"}};
    map<string, string> map_jump {
        {"null", "000"}, {"JGT", "00
        1"}, {"JEQ", "010"}, {"JGE", "011"},
        {"JLT", "100"}, {"JNE", "101"}, {"JLE", "110"}, {"JMP", "111"}};
public:
    string dest(string mnemonic) {
        return map_dest[mnemonic];
    }

    string comp(string mnemonic) {
        return map_comp[mnemonic];
    }

    string jump(string mnemonic) {
        return map_jump[mnemonic];
    }
};


class SymbolTable {
private:
    map<string, int> table;  // 連想配列を用意する。シンボル -> アドレス、として用いる。
public:
    SymbolTable() {
        // 定義済みシンボルを登録していく。
        addEntry("SP", 0);
        addEntry("LCL", 1);
        addEntry("ARG", 2);
        addEntry("THIS", 3);
        addEntry("THAT", 4);

        for (int i = 0; i < 16; ++i) {  // R0〜R15 を登録する。
            addEntry("R" + to_string(i), i);
        }

        addEntry("SCREEN", 16384);
        addEntry("KBD", 24576);
    }

    void addEntry(string symbol, int address) {
        table.emplace(symbol, address);
    }

    bool contains(string symbol) {
        return table.find(symbol) != table.end();
    }

    int getAddress(string symbol) {
        return table[symbol];
    }
};


int main(int argc, char **argv) {
    assert(argc == 2);
    char *path_input = argv[1];
    Code code;
    SymbolTable table;
    int address = 0;
    Parser *parser;
    parser = new Parser(path_input);

    // 最初のパスを行う。
    while (parser->hasMoreCommands()) {
        parser->advance();

        CommandType t = parser->commandType();

        if (t == L_COMMAND) {
            string symbol = parser->symbol();
            table.addEntry(symbol, address);
        } else if (t == A_COMMAND || t == C_COMMAND) {
            ++address;
        }
    }

    delete parser;
    parser = new Parser(path_input);
    address = 16;

    // 2 回目のパスを行う。
    while (parser->hasMoreCommands()) {
        parser->advance();
        CommandType t = parser->commandType();
        int integer;
        
        if (t == A_COMMAND) {
            string symbol = parser->symbol();
            if (isdigit(symbol[0])) {  // @Xxx が数字である。
                integer = stoi(symbol);  // 文字列->整数。
            } else {
                if (!table.contains(symbol)) {  // 未登録である。
                    table.addEntry(symbol, address);
                    ++address;
                }

                integer = table.getAddress(symbol);
            }

            // 二進数文字列を出力する
            cout << bitset<16>(integer) << endl;
        } else if (t == C_COMMAND) {
            // comp、dest、jump の各部分を二進数に変換して出力する。
            cout << "111";
            cout << code.comp(parser->comp());
            cout << code.dest(parser->dest());
            cout << code.jump(parser->jump());
            cout << endl;
        }
    }

    delete parser;
    return 0;
}
